### PR TITLE
Feature/wishlist core update

### PR DESCRIPTION
**Pull request by benRmc**
**Description**

**Issues that should be CLOSED by merge of this PR:**
* Fixes #

**Related issues that should be LINKED to this PR:**
* Connects #

### CODE REVIEW

File: traveler-swift-core/TravelerKit/AuthPath.swift [hunk 1]
```diff
@@ -12,6 +12,7 @@ enum AuthPath {
     case flights(FlightQuery)
     case catalog(CatalogQuery)
     case catalogItem(Product)
+    case catalogItemId(Product, travelerId: String)
```
Comment (ata-n) [inline traveler-swift-core/TravelerKit/AuthPath.swift]: Not sure what this is.

Naming isnt super clear

Comment (ata-n) [reply traveler-swift-core/TravelerKit/AuthPath.swift]: use the same route and make travelerId optional

File: traveler-swift-core/TravelerKit/Models/CatalogItemDetails.swift [hunk 2]
```diff
@@ -20,6 +20,7 @@ public struct CatalogItemDetails: Decodable, Product {
     public let imageUrls: [URL]
     /// Attributes
     public let information: [Attribute]?
+    public var isWishlisted: Bool
```
Comment (ata-n) [inline traveler-swift-core/TravelerKit/Models/CatalogItemDetails.swift]: I think an optional value is more suitable here.

When the user is not signed in and model says `false` can be misleading.
unless the API endpoint comes back with false. Maybe we should discuss with backend what is most appropriate

Comment (ata-n) [reply traveler-swift-core/TravelerKit/Models/CatalogItemDetails.swift]: Documentation on public properties on functions

File: traveler-swift-core/TravelerKit/Traveler.swift [hunk 3]
```diff
@@ -235,6 +240,95 @@ public class Traveler {
...
+        var immediateResult = result
+        if let result = result, let items = items as? [CatalogItem] {
+            let allItemsExistInResult = items.map({ $0.id }).reduce(true) { (exists, id) -> Bool in
+                exists && result.items.values.contains(where: { (item) -> Bool in
+                    item.id == id
+                })
+            }
+
+            if !allItemsExistInResult {
+                completion(nil, WishlistResultError.itemsDontExistInResult)
```
Comment (ata-n) [inline traveler-swift-core/TravelerKit/Traveler.swift]: Wrong error type used here

theres a WishlistToggleError type that is public.

Maybe provide better naming there as well

File: traveler-swift-core/TravelerKit/Traveler.swift [hunk 4]
```diff
@@ -235,6 +240,95 @@ public class Traveler {
...
+
+        var immediateResult = result
+        if let result = result, let items = items as? [CatalogItem] {
+            let allItemsExistInResult = items.map({ $0.id }).reduce(true) { (exists, id) -> Bool in
+                exists && result.items.values.contains(where: { (item) -> Bool in
+                    item.id == id
+                })
+            }
+
+            if !allItemsExistInResult {
```
Comment (ata-n) [inline traveler-swift-core/TravelerKit/Traveler.swift]: use guard statements

File: traveler-swift-core/TravelerKit/Traveler.swift [hunk 5]
```diff
@@ -235,6 +240,95 @@ public class Traveler {
...
+                completion(nil, WishlistResultError.itemsDontExistInResult)
+                return nil
+            }
+
+            for item in items {
+                immediateResult!.remove(item)
```
Comment (ata-n) [inline traveler-swift-core/TravelerKit/Traveler.swift]: no need to force unwrap here...use `if let` statement for your immediate resuts

File: traveler-swift-core/TravelerKit/Models/WishlistToggleError.swift [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+import Foundation
+
+public enum WishlistToggleError: Error {
+    case unidentifiedTraveler
+    case unableToAdd
+    case unableToRemove
+    case itemAlreadyInWishlist
+    case notInWishlist
+}
```
Comment (ata-n) [inline traveler-swift-core/TravelerKit/Models/WishlistToggleError.swift]: remove any unused and unwanted error types and cases

File: traveler-swift-core/TravelerKit/Models/CatalogItemDetails.swift (hunk 2 above)
Comment (benRmc) [reply traveler-swift-core/TravelerKit/Models/CatalogItemDetails.swift]: Currently the API returns 'false' when user is not logged in.

File: traveler-swift-core/TravelerKit/Traveler.swift (hunk 5 above)
Comment (benRmc) [reply traveler-swift-core/TravelerKit/Traveler.swift]: Here we can guarantee that that the immediateResult is not nil.

File: traveler-swift-core/TravelerKit/Models/CatalogItemDetails.swift (hunk 2 above)
Comment (ata-n) [reply traveler-swift-core/TravelerKit/Models/CatalogItemDetails.swift]: This should be optional

File: traveler-swift-core/TravelerKit/Delegates/WishlistAddDelegate.swift [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+//  Created by Omar Padierna on 2019-09-03.
+//  Copyright © 2019 Ata Namvari. All rights reserved.
+//
+
+import Foundation
+
+/// Notified when wishlisting item
+public protocol WishlistAddDelegate: class {
+    /**
+     Called when the `CatalogItem` was wishlisted successfully
```
Comment (Alvtag) [inline traveler-swift-core/TravelerKit/Delegates/WishlistAddDelegate.swift]: add documentation

**Pull request merged**